src/type/schema.ts

```diff
@@ -427,33 +428,46 @@ export interface GraphQLSchemaNormalizedConfig extends GraphQLSchemaConfig {
 }
 
 function collectReferencedTypes(
-  type: GraphQLType,
+  root: GraphQLNamedType,
   typeSet: Set<GraphQLNamedType>,
 ): Set<GraphQLNamedType> {
-  const namedType = getNamedType(type);
+  const typesToCheck: Array<GraphQLNamedType> = [root];
+
+  while (typesToCheck.length) {
+    const namedType = typesToCheck.pop();
+
+    if (!namedType || typeSet.has(namedType)) {
+      continue;
+    }
 
-  if (!typeSet.has(namedType)) {
     typeSet.add(namedType);
+
+    const typesToAdd: Array<GraphQLNamedType> = [];
+
     if (isUnionType(namedType)) {
       for (const memberType of namedType.getTypes()) {
-        collectReferencedTypes(memberType, typeSet);
+        typesToAdd.push(getNamedType(memberType));
       }
     } else if (isObjectType(namedType) || isInterfaceType(namedType)) {
       for (const interfaceType of namedType.getInterfaces()) {
-        collectReferencedTypes(interfaceType, typeSet);
+        typesToAdd.push(interfaceType);
       }
 
       for (const field of Object.values(namedType.getFields())) {
-        collectReferencedTypes(field.type, typeSet);
+        typesToAdd.push(getNamedType(field.type));
+
         for (const arg of field.args) {
-          collectReferencedTypes(arg.type, typeSet);
+          typesToAdd.push(getNamedType(arg.type));
         }
       }
     } else if (isInputObjectType(namedType)) {
       for (const field of Object.values(namedType.getFields())) {
-        collectReferencedTypes(field.type, typeSet);
+        typesToAdd.push(getNamedType(field.type));
       }
     }
+
+    typesToAdd.reverse()
```

Comment:
There are tests that ensure types are found in a specific order which is a depth-first search, this maintains compatibility with that order and implements a depth-first search with iteration. As there were tests for it I assume the order is somewhat important, but its not clear why exactly.